File: lib/src/models/nodes/node_config.dart
import 'package:flutter/material.dart';
import 'package:flutter_sceneview/src/models/materials/base_material.dart';
import 'package:flutter_sceneview/src/models/shapes/base_shape.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'node_config.freezed.dart';
part 'node_config.g.dart';

@Freezed(unionKey: 'type')
@freezed
abstract class NodeConfig with _$NodeConfig {
  @JsonSerializable(explicitToJson: true)
  // if model is not empty, then the load default flag should be set
  // automatically
  const factory NodeConfig.model({
    String? fileName,
    @Default(false) bool loadDefault,
  }) = ModelConfig;

  @JsonSerializable(explicitToJson: true)
  const factory NodeConfig.shape({
    @BaseShapeConverter() BaseShape? shape,
    @BaseMaterialConverter() BaseMaterial? material,
  }) = ShapeConfig;

  @JsonSerializable(explicitToJson: true)
  const factory NodeConfig.text({
    required String text,
    String? fontFamily,
    @ColorConverter()
    @Default(Colors.white)
    Color textColor, // Default to white
    @Default(1.0) double size,
  }) = TextConfig;

  const factory NodeConfig.anchor({String? anchorId}) = AnchorConfig;

  const factory NodeConfig.unknown() = UnknownConfig;

  factory NodeConfig.fromJson(Map<String, dynamic> json) =>
      _$NodeConfigFromJson(json);
}

class NodeConfigConverter
    extends JsonConverter<NodeConfig, Map<Object?, Object?>> {
  const NodeConfigConverter();

  @override
  NodeConfig fromJson(Object? json) {
    if (json == null) {
      throw ArgumentError('Expected non-null JSON for NodeConfig');
    }
    if (json is! Map) {
      throw ArgumentError(
        'Expected a Map for NodeConfig but got ${json.runtimeType}',
      );
    }

    return NodeConfig.fromJson(Map<String, dynamic>.from(json));
  }

  @override
  Map<String, dynamic> toJson(NodeConfig object) => object.toJson();
}

File: android/src/main/kotlin/com/example/flutter_sceneview/models/nodes/NodeConfig.kt
package com.example.flutter_sceneview.models.nodes

import com.example.flutter_sceneview.models.materials.BaseMaterial
import com.example.flutter_sceneview.models.shapes.BaseShape

enum class NodeType {
    MODEL, SHAPE, TEXT, ANCHOR, UNKNOWN
}

sealed class NodeConfig

data class ModelConfig(val fileName: String? = null, val loadDefault: Boolean? = false) :
    NodeConfig()

data class ShapeConfig(val shape: BaseShape? = null, val material: BaseMaterial? = null) :
    NodeConfig()

data class TextConfig(
    val text: String? = null,
    val fontFamily: String? = null,
    val textColor: Int = BaseMaterial.DEFAULT.color,
    val size: Float = 1f // Width in meters
) : NodeConfig()